File: app/src/main/java/com/mapelli/simone/githubclient/data/GithubUsersDatabase.java
package com.mapelli.simone.githubclient.data;

import android.content.Context;
import android.util.Log;

import com.mapelli.simone.githubclient.data.entity.UserProfile_Full;
import com.mapelli.simone.githubclient.data.entity.UserProfile_Mini;
import com.mapelli.simone.githubclient.data.entity.UserRepository;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import androidx.room.TypeConverters;

@Database(entities = {UserRepository.class, UserProfile_Full.class, UserProfile_Mini.class},
        version = 1, exportSchema = false)
@TypeConverters(DateConverter.class)
public abstract class GithubUsersDatabase extends RoomDatabase {
    private static final String TAG = GithubUsersDatabase.class.getSimpleName();

    // lock for synchro
    private static final Object LOCK   = new Object();
    private static final String DBNAME = "GithubUsersDB";
    private static GithubUsersDatabase sDbInstance ;

    // get for the dao
    public abstract GithubUsersDbDao githubUsersDao();


    public static GithubUsersDatabase getsDbInstance(Context context){
        if(sDbInstance == null){
            synchronized (LOCK){
                Log.d(TAG, "Creating App db singleton instance...");
                sDbInstance = Room.databaseBuilder(context.getApplicationContext(), GithubUsersDatabase.class,GithubUsersDatabase.DBNAME)
                        //.allowMainThreadQueries() // TODO : temporary in case of debugging, delete this
                        .build();
            }

        }
        Log.d(TAG, "Db created");
        return sDbInstance;
    }


}

File: app/src/main/java/com/mapelli/simone/githubclient/data/GithubUsersDatabase.kt
package com.mapelli.simone.githubclient.data

import android.content.Context
import android.util.Log

import com.mapelli.simone.githubclient.data.entity.UserProfile_Full
import com.mapelli.simone.githubclient.data.entity.UserProfile_Mini
import com.mapelli.simone.githubclient.data.entity.UserRepository

import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.room.TypeConverters

@Database(entities = [UserRepository::class, UserProfile_Full::class, UserProfile_Mini::class],
        version = 1, exportSchema = false)
@TypeConverters(DateConverter::class)
abstract class GithubUsersDatabase : RoomDatabase() {

    abstract fun githubUsersDao(): GithubUsersDbDao

    companion object {
        private val TAG = GithubUsersDatabase::class.java.getSimpleName()

        // lock for synchro
        private val LOCK = Any()
        private val DBNAME = "GithubUsersDB"
        @Volatile private var sDbInstance: GithubUsersDatabase? = null


        /*
        fun getsDbInstance(context: Context): GithubUsersDatabase {
            return sDbInstance ?: synchronized(this) {
                sDbInstance ?: buildDb(context).also { sDbInstance = it }
            }
        }

        private fun buildDb(context: Context) : GithubUsersDatabase{
            return Room.databaseBuilder<GithubUsersDatabase>(context.applicationContext,
                    GithubUsersDatabase::class.java, DBNAME)
                    .build()
        }
        */

        // BEFORE :
        fun getsDbInstance(context: Context): GithubUsersDatabase? {
            if (sDbInstance == null) {
                synchronized(this) {
                    Log.d(TAG, "Creating App db singleton instance...")
                    sDbInstance = Room.databaseBuilder<GithubUsersDatabase>(context.applicationContext,
                            GithubUsersDatabase::class.java, GithubUsersDatabase.DBNAME)
                            .build()
                }

            }
            Log.d(TAG, "Db created")
            return sDbInstance
        }


    }


}
